### PR TITLE
Build SYCL kernels for ATen XPU ops on Native Windows (take 2)

### PR DESCRIPTION
**Pull request by min-jean-cho**
Original PR https://github.com/pytorch/pytorch/pull/126725 is closed due to bad rebase.

-------
As proposed in https://github.com/pytorch/pytorch/issues/126719, we are enabling PyTorch XPU on Native Windows on Intel GPU. 

This PR  enables XPU build on Windows as the first step of #126719:

- Enable `USE_XPU` build on Windows using MSVC as host compiler. The use of MSVC as host compiler seamlessly aligns with the existing PyTorch build on Windows.
- Build oneDNN GPU library on Windows.

cc @peterjc123 @mszhanyi @skyline75489 @nbcsm @vladimir-aubrecht @iremyux @Blackhex @cristianPanaite @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @PenghuiCheng @jianyuh @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @EikanWang @fengyuan14 @guangyey

### CODE REVIEW

File: cmake/Modules/FindMKLDNN.cmake [hunk 1]
```diff
@@ -21,10 +21,16 @@ IF(NOT MKLDNN_FOUND)
...
-      set(DNNL_HOST_COMPILER "g++")
+      # Linux
       # g++ is soft linked to /usr/bin/cxx, oneDNN would not treat it as an absolute path
+      set(DNNL_HOST_COMPILER "g++")
+      set(SYCL_CXX_DRIVER "icpx")
+      set(DNNL_LIB_NAME "libdnnl.a")
     else()
-      message(FATAL_ERROR "oneDNN library currently only supports GUN g++ compiler for XPU backend")
+      # Windows
+      set(DNNL_HOST_COMPILER "DEFAULT")
```
Comment (min-jean-cho) [inline cmake/Modules/FindMKLDNN.cmake:31]: @EikanWang @mdtoguchi @vpirogov, let's move https://github.com/pytorch/pytorch/pull/126725/files#r1618284937 discussion to here.

File: cmake/Modules/FindSYCLToolkit.cmake [hunk 2]
```diff
@@ -49,12 +49,27 @@ find_file(
...
-find_library(
-  SYCL_LIBRARY
-  NAMES sycl
-  HINTS ${SYCL_LIBRARY_DIR}
-  NO_DEFAULT_PATH
-)
+# On both Linux and Windows, the current version of SYCL runtime is 7.
+# On Linux, libsycl.so symbolic links to libsycl.so.7.
+# On Windows, sycl7.dll version suffix is used.
+foreach(sycl_lib_version "" 7)
```
Comment (gujinghui) [inline cmake/Modules/FindSYCLToolkit.cmake:55]: Pls submit a JIRA to track the requirement to oneAPI.

Comment (min-jean-cho) [reply cmake/Modules/FindSYCLToolkit.cmake:55]: Thanks, will submit. Future reviewer, see https://github.com/pytorch/pytorch/pull/126725#discussion_r1616909975 for background.

File: cmake/Modules/FindMKLDNN.cmake (hunk 1 above)
Comment (min-jean-cho) [reply cmake/Modules/FindMKLDNN.cmake:31]: > Generated binaries from icx, icpx, icx-cl on Windows will always be MSVC compatible unless there is an override via the --target option. 

Based on @mdtoguchi https://github.com/pytorch/pytorch/pull/126725#discussion_r1619027194 above, I suppose the compatibility concern of `icx` (clang under the hood) and MSVC is not an issue now with the use of `ONEDNN_DPCPP_HOST_COMPILR=DEFAULT`.

I wonder if it's the same case on Linux with `icpx` (clang under the hood) and g++. And if so, what was the original need to use `ONEDNN_DPCPP_HOST_COMPILER=g++` on Linux?

Comment (vpirogov) [reply cmake/Modules/FindMKLDNN.cmake:31]: If I recall correctly the reason to use g++ in Linux builds was threading. Pytorch relies on GNU OpenMP library and the intent was to generate GNU OpenMP code for oneDNN threading on CPU.

Comment (EikanWang) [reply cmake/Modules/FindMKLDNN.cmake:31]: > Generated binaries from icx, icpx, icx-cl on Windows will always be MSVC compatible unless there is an override via the --target option. Use of icx or icx-cl is recommended if you require MSVC compatible command line options as well. icpx will use Linux based options, but will target MSVC when run on Windows.

Sounds good. By the way, does `--target` here means that users override the `--target` and then impact the oneDNN? @mdtoguchi

Comment (mdtoguchi) [reply cmake/Modules/FindMKLDNN.cmake:31]: I wouldn't expect users to override the `--target` values.  The target values allow for modifying the target architecture (say x86_64 vs powerpc for example) as well as changing the OS (Windows vs Linux) or vendor (gnu vs msvc) settings.  The use of this option is useful especially in cross-compilation environments, but is not expected to be used in typical compilation scenarios.  Any usage of this option will also require the dependent libraries to be built with that same target expectation.

File: cmake/Modules/FindSYCLToolkit.cmake (hunk 2 above)
Comment (jgong5) [reply cmake/Modules/FindSYCLToolkit.cmake:55]: If it is Windows specific limitation, I would suggest we keep the code for Linux unchanged. Add comments on why Windows is special and what we plan to do in the future.

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: cmake/Modules/FindSYCLToolkit.cmake (hunk 2 above)
Comment (min-jean-cho) [reply cmake/Modules/FindSYCLToolkit.cmake]: Thanks, updated. It is Windows specific limitation.

**Review by guangyey — APPROVED** [review]

**Review by EikanWang — APPROVED** [review]

**Comment by min-jean-cho** [comment]
Hi @ezyang , @albanD, this PR adds XPU support on Native Windows for Intel GPUs. Would you please help review from Meta side? Thanks.

File: aten/src/ATen/xpu/detail/XPUHooks.cpp [hunk 3]
```diff
@@ -25,7 +25,13 @@ std::string XPUHooks::showConfig() const {
 
 int32_t XPUHooks::getGlobalIdxFromDevice(const at::Device& device) const {
   TORCH_CHECK(device.is_xpu(), "Only the XPU device type is expected.");
+#ifdef _WIN32
+  TORCH_CHECK(
+      false,
+      "Default context is not supported on XPU on Windows. So we can NOT find its global index of the ATen device.");
```
Comment (albanD) [inline aten/src/ATen/xpu/detail/XPUHooks.cpp:31]: What is using this function? What is the impact of forbidding this?

File: c10/xpu/XPUMacros.h [hunk 4]
```diff
@@ -1,15 +1,29 @@
...
+#include <c10/xpu/impl/xpu_cmake_macros.h>
+#endif
+
 // See c10/macros/Export.h for a detailed explanation of what the function
 // of these macros are.  We need one set of macros for every separate library
 // we build.
 
+#ifdef _WIN32
+#if defined(C10_XPU_BUILD_SHARED_LIBS)
+#define C10_XPU_EXPORT __declspec(dllexport)
```
Comment (albanD) [inline c10/xpu/XPUMacros.h:13]: Why can't you use the regular C10_EXPORT / C10_API here?

File: torch/csrc/xpu/Module.cpp [hunk 5]
```diff
@@ -11,24 +11,30 @@
 #include <torch/csrc/utils/python_numbers.h>
 #include <torch/csrc/utils/python_strings.h>
 
+#ifndef WIN32
```
Comment (albanD) [inline torch/csrc/xpu/Module.cpp:14]: Wasn't this copied from the cuda version? Why is this needed here but not there?

**Review by albanD — COMMENTED** [review]
Will let @malfet review the detailed build change

File: torch/csrc/xpu/Module.cpp (hunk 5 above)
Comment (min-jean-cho) [reply torch/csrc/xpu/Module.cpp:14]: `pthread.h` is not supported on Windows. It's the same with other backends:
https://github.com/pytorch/pytorch/blob/b704c7cf0f8fb8b92198ffd4cd07716a7b9efae1/torch/csrc/cuda/Module.cpp#L53-L55

https://github.com/pytorch/pytorch/blob/b704c7cf0f8fb8b92198ffd4cd07716a7b9efae1/torch/csrc/mtia/Module.cpp#L11-L13

File: c10/xpu/XPUMacros.h (hunk 4 above)
Comment (min-jean-cho) [reply c10/xpu/XPUMacros.h:13]: No particular reason besides alignment with other backends, we can use `C10_EXPORT`. 
Same for Linux, we can use `C10_EXPORT` instead of `#define C10_XPU_EXPORT __attribute__((__visibility__("default")))`.

File: aten/src/ATen/xpu/detail/XPUHooks.cpp (hunk 3 above)
Comment (guangyey) [reply aten/src/ATen/xpu/detail/XPUHooks.cpp:31]: This function is mainly used for dlpack. We want to inform users that dlpack doesn't support XPU on Windows, since we can't ensure context consistency because the default context is not being used on Windows for XPU.

**Review by gujinghui — APPROVED** [review]

**Review by ezyang — APPROVED** [review]
seems fine ig

**Comment by min-jean-cho** [comment]
The failures seem unrelated.

**Comment by guangyey** [comment]
@pytorchbot merge -i

The failures are unrelated.
